Improve ywasm performance by 2x (ymap_set case)

MarkJGx opened the pull request:
This change set contains various optimizations to ywasm.
- Optimize hot path calls that end up crossing the JavaScript boundary. (Checkout: https://hacks.mozilla.org/2018/10/calls-between-javascript-and-webassembly-are-finally-fast-%F0%9F%8E%89/)
- Use talc instead of dlmalloc for global heap allocations. Negligible imporvement, but faster is faster.

MarkJGx committed:
ywasm: replace get_type with direct js function to avoid reflection overhead

MarkJGx committed:
ywasm: get rid of expensive string allocs in mut_from_js

MarkJGx committed:
ywasm: Cargo: match wasm-bindgen version from root

MarkJGx committed:
ywasm: use talc alloactor (test)

MarkJGx committed:
ywasm: transaction: use faster pointer lookup function

MarkJGx committed:
ywasm: avoid string alloc if not a stringg on as_value, check string after basics

MarkJGx committed:
ywasm: as_value: pass through javascript with...

rust closure to avoid unnecessary call overhead

MarkJGx committed:
ywasm:test with no ptr get string alloc

MarkJGx committed:
ywasm: handle all result cases in as_value

MarkJGx committed:
ywasm: reflection remove static strings as they will remain unusued for now

MarkJGx committed:
Cargo: remove spin

MarkJGx committed:
ywasm: Js: address fixme

MarkJGx committed:
ywasm: fix bool usage

MarkJGx committed:
ywasm: Reflection: re-introduce JS_PTR immutable

MarkJGx committed:
talc: specify version

MarkJGx committed:
Cargo.lock: remove unusued dependencies

katzenbar reviewed (approved):
Notes from discussion

- Still needs a cleanup pass before we would want to upstream, to make the value type enums more readable.
- This requires nightly to build
- The external JS module doesn't work with the nodejs build in wasm-pack (see note on https://rustwasm.github.io/wasm-bindgen/reference/js-snippets.html), which breaks the tests

MarkJGx committed:
DeleteSet: optimize bulk squash for improved performance

- Significantly improves performance during GC by batching squashes.
- Internal benchmarks with ~34,000 GC block cells reduced runtime from 118s to 400ms (~300x speedup) in release builds.

MarkJGx committed:
Merge pull request #3 from Generate-Technologies/markjg/cloud/gc/optimization-refined

yrs: DeleteSet: optimize bulk squash for improved performance
